File: Python_Solutions/Ransom_Note.py
"""
Leetcode Challenge: Ransom Note

Given two strings ransomNote and magazine, return true if ransomNote can be constructed by using the letters from magazine and false otherwise.

Each letter in magazine can only be used once in ransomNote.

4/30/24
Mark Robuck
"""
class Solution:
    """
    Determine if all characters in a string are contained within another string.
    Args:
        ransomNote(str): the string to be tested if all character are also within magazine
        magazine(str): the string to be used to test if all characters from ransomNote are within it
    Returns:
        bool: returns true if all characters in ransomNote are within magazine, and false if not
    """
    def canConstruct(self, ransomNote: str, magazine: str) -> bool:
        containsAll = True
        #create an array for each string containing each letter using split
        ransomLetters = list(ransomNote)
        magazineLetters = list(magazine)
        #loop through ransom array
        for i in ransomLetters:
            #check to see if each letter in ransom is in magazine
            if i in magazineLetters:
                #if yes remove it from magazine
                magazineLetters.remove(i)
                #if reach end of loop return true
            else:
            #if check is false return false
                containsAll = False
        return containsAll

#An interesting side note I found after reviewing some of the other solutions submitted.  Removing my flag variable and replacing lines 16 and 17 with return true/false yields a runtime of half the time my solution ran in.

File: Java_Solutions/Ransom_Note.java
/*
* Leetcode Challenge: Ransom Note
* 
* Given two strings ransomNote and magazine, return true if ransomNote can be constructed by using the letters from magazine and false otherwise.
* 
* Each letter in magazine can only be used once in ransomNote.
* 
* 4/30/24
* Mark Robuck
*/
import java.util.ArrayList;

class Solution {
    /*
    * Determine if all characters in a string are contained within another string.
    * @params ransomNote, the string to be tested if all character are also within magazine
    * @params magazine, the string to be used to test if all characters from ransomNote are within it
    * @return true if all characters in ransomNote are within magazine, and false if not
    */
    public boolean canConstruct(String ransomNote, String magazine) {
        boolean containsAll = true;
        //create an array of letters for each string
        ArrayList<Character> ransomLetters = new ArrayList<Character>();
        ArrayList<Character> magazineLetters = new ArrayList<Character>();
        for (char c : ransomNote.toCharArray()) {
            ransomLetters.add(c);
        }
        for (char c : magazine.toCharArray()) {
            magazineLetters.add(c);
        }
        //loop through ransom array
        for (char c: ransomLetters) {
            //check to see if each letter in ransom is in magazine
            if (magazineLetters.contains(c)) {
                //if yes remove it from magazine
                    magazineLetters.remove(Character.valueOf(c));
            }
            //else if letter not in magazine return false
            else{
                return false;
            }
        }
        //if reach the end of the loop return true
        return true;
    }
}
